### PR TITLE
tests: drivers: flash: spi_flash: add testcode for the OSPI flash

### PR DESCRIPTION
**Pull request by Saba-AlifSemi**
Add modular test suite for OSPI flash driver with basic, boundary, negative, performance, and XIP test categories.

### CODE REVIEW

File: tests/drivers/flash/spi_flash/src/test_ospi_main.c [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+	for (i = 0; i < BUFF_SIZE; i++) {
+		if (r_buf[i] != flash_param->erase_value) {
+			count++;
+		}
+	}
+
+	LOG_INF("Total errors after reading erased chip = %d", count);
+}
+
+static void multi_page_test(const struct device *dev)
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_main.c:234]: This test is doing multi-page means, y its doing sector erase, only writing offset write once .

File: tests/drivers/flash/spi_flash/src/test_ospi_main.c [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+	/* Test 1: All zeros pattern */
+	LOG_INF("Test all-zeros pattern");
+	memset(w_buf, 0x00, BUFF_SIZE);
+
+	ret = flash_erase(flash_dev,
+			  SPI_FLASH_TEST_REGION_OFFSET,
+			  SPI_FLASH_SECTOR_SIZE);
+	zassert_equal(ret, 0, "Erase failed [%d]", ret);
+
+	ret = flash_write(flash_dev,
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_main.c]: Here, instead of writing, as always, you need to fill the same pattern using the flash_fill() function, which would be helpful

File: tests/drivers/flash/spi_flash/src/test_ospi_negative_tests.c [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/ztest.h>
+
+#include "test_ospi_flash_test.h"
+
+LOG_MODULE_REGISTER(ospi_negative, LOG_LEVEL_INF);
+
+/* -------- Test: write without erase (negative scenario) -------- */
+static void write_without_erase_test(void)
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_negative_tests.c:23]: This test, I think, needs to be enhanced because only sector write and sector read without erase is valid if the test is run with other tests again, it needs something as you read first, then write, then you need to validate the second read is different from the first read or something like this. Let me know any ideas you have.

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	LOG_INF("=== Boundary Size Operations Test ===");
+
+	/* Erase test region */
+	ret = flash_erase(flash_dev, SPI_FLASH_TEST_REGION_OFFSET,
+			  SPI_FLASH_SECTOR_SIZE);
+	zassert_equal(ret, 0, "Erase failed [%d]", ret);
+
+	/* Test minimum write size (write_block_size) */
+	memset(min_buf, 0xAA, sizeof(min_buf));
+	ret = flash_write(flash_dev, SPI_FLASH_TEST_REGION_OFFSET,
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c]: flash_fill() Can we use this function?

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+		zassert_equal(e_count, 0,
+			     "Verify failed at 0x%x: %d mismatches",
+			     offset, e_count);
+	}
+
+	LOG_INF("Address boundary test PASSED");
+}
+
+/* -------- Test: boundary size operations -------- */
+static void boundary_size_operations_test(void)
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c:139]: smells similar to the data pattern test.

**Comment by sudiiyengar-en** [comment]
I see that there are still some api we can make use of, like flash_page_foreach(), flash_flatten(), like this.
Use of LOG should be for debugging and to pass some info; if test output should print something, use TC_PRINT.

**Review by sudiiyengar-en — CHANGES_REQUESTED** [review]

File: tests/drivers/flash/spi_flash/README.rst [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+Manual Build Instructions
+*************************
+
+**Basic build using snippet system (recommended):**
+
+.. code-block:: console
+
+   west build -p always -b alif_e7_dk/ae722f80f55d5xx/rtss_he \\
+     tests/drivers/spi_flash/ -S ospi-flash
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/README.rst]: Where are the snippets? Please review before the code push.

File: tests/drivers/flash/spi_flash/src/test_ospi_xip_tests.c [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+static void *test_ospi_xip_setup(void)
+{
+	LOG_INF("=== OSPI XIP Test Suite Setup ===");
+	ospi_flash_setup_device();
+	return NULL;
+}
+
+static void test_ospi_xip_before(void *fixture)
+{
```
Comment (Shamshoor-AlifSemi) [inline tests/drivers/flash/spi_flash/src/test_ospi_xip_tests.c:87]: test_ospi_xip_setup and test_ospi_xip_before currently perform identical initialization. Since setup runs once per suite and before runs per test, is this duplication intentional for future test cases? If not, we could consolidate or differentiate their responsibilities.

**Review by Shamshoor-AlifSemi — CHANGES_REQUESTED** [review]

File: tests/drivers/flash/spi_flash/README.rst (hunk 6 above)
Comment (Saba-AlifSemi) [reply tests/drivers/flash/spi_flash/README.rst]: Using the same snippets as the sample app; they are available under snippets.
The build command here reuses those existing snippet definitions, so no new snippet set is introduced in this test README.

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c (hunk 4 above)
Comment (Saba-AlifSemi) [reply tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c]: Yes, updated. Since these are uniform byte patterns, I switched the boundary-size writes to flash_fill() (for both write-block-size and page-size cases), while keeping readback verification unchanged.

**Review by VeijoPesonen — APPROVED** [review]

**Review by VeijoPesonen — DISMISSED** [review]

**Review by Shamshoor-AlifSemi — DISMISSED** [review]
looks good.

File: tests/drivers/flash/spi_flash/src/test_ospi_negative_tests.c [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+		}
+	}
+
+	LOG_INF("Unaligned address test PASSED");
+}
+
+/* ======== Negative Test Suite ======== */
+
+/* TESTLINK TEST CASE - ZTC-2008 */
+ZTEST(test_ospi_negative, test_ospi_write_without_erase)
```
Comment (MaheshYerra-AlifSemi) [inline tests/drivers/flash/spi_flash/src/test_ospi_negative_tests.c:356]: please remove testcase numbers

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+/* ======== Boundary Test Suite ======== */
+
+/* TESTLINK TEST CASE - ZTC-2001 */
+ZTEST(test_ospi_boundary, test_ospi_address_boundaries)
+{
+	address_boundary_test();
+}
+
+/* TESTLINK TEST CASE - ZTC-2002 */
+ZTEST(test_ospi_boundary, test_ospi_cross_page_boundary)
```
Comment (MaheshYerra-AlifSemi) [inline tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c:227]: /* TESTLINK TEST CASE - ZTC-2001 */ please remove testcase numbers

**Review by MaheshYerra-AlifSemi — DISMISSED** [review]
changes required

**Review by VeijoPesonen — DISMISSED** [review]

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+	LOG_INF("=== Cross-Page Boundary Test ===");
+
+	/* Prepare test data */
+	for (int i = 0; i < BUFF_SIZE; i++) {
+		w_buf[i] = (i + 100) % 256;
+	}
+
+	/* Use an offset that crosses a page boundary based on device page size */
+	cross_offset = SPI_FLASH_TEST_REGION_OFFSET +
+		       flash_param->page_size - (BUFF_SIZE / 2);
```
Comment (sudiiyengar-en) [inline tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c:47]: /* This test assumes page_size >= BUFF_SIZE/2 so cross_offset
 * stays within the erased region starting at TEST_REGION_OFFSET.
 */
zassert_true(flash_param->page_size >= (BUFF_SIZE / 2),
	     "page_size (%zu) too small for cross-page test",
	     flash_param->page_size);

**Review by sudiiyengar-en — CHANGES_REQUESTED** [review]

**Review by VeijoPesonen — APPROVED** [review]

**Review by sudiiyengar-en — APPROVED** [review]
Looks good

**Review by Shamshoor-AlifSemi — APPROVED** [review]

File: tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c (hunk 9 above)
Comment (VeijoPesonen) [reply tests/drivers/flash/spi_flash/src/test_ospi_boundary_tests.c:227]: Seems like these have been fixed.
